lib/tasks/book2.rake

```diff
@@ -79,7 +79,8 @@ def genbook (code, &get_content)
   alldoc = Nokogiri::HTML(html)
   number = 1
 
-  book = Book.where(:edition => 2, :code => code).first_or_create
+  book = Book.where(:edition => 2, :code => code).destroy_all
+  book = Book.create(:edition => 2, :code => code)
 
```

Comment:
IIUC, any exception in the enclosing transaction block cancels the transaction, so the book can not be left in a state where it was deleted and not repopulated.

Reply:
Right, that's the theory. I didn't actually test with an exception, but I did confirm that during the rebuild we still see the old content, and then it cuts over to the new content atomically.

Reply:
Just for fun, I tested with a fake exception, and it does indeed work as expected.